core/src/services/chainsafe/lister.rs

```diff
@@ -0,0 +1,80 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use bytes::Buf;
+use http::StatusCode;
+
+use super::core::{parse_info, ChainsafeCore, Info};
+use super::error::parse_error;
+use crate::raw::oio::Entry;
+use crate::raw::*;
+use crate::*;
+
+pub struct ChainsafeLister {
+    core: Arc<ChainsafeCore>,
+    path: String,
+}
+
+impl ChainsafeLister {
+    pub fn new(core: Arc<ChainsafeCore>, path: &str) -> Self {
+        Self {
+            core,
+
+            path: path.to_string(),
+        }
+    }
+}
+
+#[async_trait]
+impl oio::PageList for ChainsafeLister {
+    async fn next_page(&self, ctx: &mut oio::PageContext) -> Result<()> {
+        let resp = self.core.list_objects(&self.path).await?;
+
+        match resp.status() {
+            StatusCode::OK => {
+                let bs = resp.into_body().bytes().await?;
+
+                let output: Vec<Info> =
+                    serde_json::from_reader(bs.reader()).map_err(new_json_deserialize_error)?;
+
+                for info in output {
+                    let mut path = if self.path.starts_with('/') {
```

Comment:
Why we need to check `self.path` like this?

Reply:
Because build_abs_path need root be start with /

Reply:
I didn't get this. All `path` input from `list` has been normalized, so it must be start with `/`.

Reply:
I replace it:
```rust
   // let mut path = if self.path.starts_with('/') {
    //     build_abs_path(&self.path, &info.name)
    // } else {
    //     build_abs_path(&format!("/{}", &self.path), &info.name)
    // };
    let mut path = build_abs_path(&self.path, &info.name);
```
has this error:
```shell
---- behavior::test_list_with_recursive ----
test panicked: root must start with /
```

Reply:
> Perhaps I should call `normalize_root` during initialization?

Yep.

Reply:
By the way, `build_abs_path` is designed to play with `root`. It's undefined behavior when you put a `path` in it directly. I'm fine to use `format!("{path}/name").trim_start_matches('/')` here.

Reply:
I understand that path and root mean the same thing in this scenario

Reply:
> I understand that path and root mean the same thing in this scenario

The term `root` refers to the base of this service, which has been standardized by `normalize_root`. However, it's not a significant problem. Address it using your preferred methods.

Reply:
Ok, this also fixed